segments_tree.go

```diff
@@ -0,0 +1,141 @@
+package quamina
+
+import (
+	"fmt"
+	"strings"
+)
+
+const SegmentSeparator = "\n"
+
+type segmentsTree struct {
+	root bool
+
+	nodes map[string]*segmentsTree
+
+	// fields map segment to full path
+	// will be present only on the leafs
+	fields map[string][]byte
+}
+
+func newSegmentsIndex(paths ...string) *segmentsTree {
+	st := newSegmentsIndexNode(true)
+
+	for _, path := range paths {
+		st.add(path)
+	}
+
+	return st
+}
+
+func newSegmentsIndexNode(root bool) *segmentsTree {
+	return &segmentsTree{
+		root:   root,
+		nodes:  make(map[string]*segmentsTree),
+		fields: make(map[string][]byte),
+	}
+}
+
+func (p *segmentsTree) add(path string) {
+	parts := strings.Split(path, SegmentSeparator)
+
+	if len(parts) == 1 {
+		// It's a direct field.
+		p.fields[path] = []byte(path)
+		return
+	}
+
+	var node *segmentsTree
+
+	node = p.getOrCreate(parts[0])
+
+	for i, part := range parts[1:] {
+		if i == len(parts)-2 {
+			node.addField(part, []byte(path))
+		} else {
+			node = node.getOrCreate(part)
+		}
+	}
+}
+
+func (p *segmentsTree) getOrCreate(name string) *segmentsTree {
+	_, ok := p.nodes[name]
+	if !ok {
+		p.nodes[name] = newSegmentsIndexNode(false)
+	}
+
+	return p.nodes[name]
+}
+
+func (p *segmentsTree) addField(name string, path []byte) {
+	_, ok := p.fields[name]
+	if !ok {
+		p.fields[name] = path
+	}
+}
+
+// Get implements SegmentsTreeTracker
+func (p *segmentsTree) Get(name []byte) (SegmentsTreeTracker, bool) {
+	n, ok := p.nodes[string(name)]
+	return n, ok
+}
+
+// IsRoot implements SegmentsTreeTracker
+func (p *segmentsTree) IsRoot() bool {
+	return p.root
+}
+
+// IsSegmentUsed implements SegmentsTreeTracker
+func (p *segmentsTree) IsSegmentUsed(segment []byte) bool {
```

Comment:
Another place that's hard to understand, could use a little commenting.

Reply:
added comment